src/librustc_typeck/collect.rs

```diff
@@ -1020,63 +1019,63 @@ fn convert_union_def<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>,
-        let print_err = |cv: ConstVal| {
-            struct_span_err!(ccx.tcx.sess, e.span, E0079, "mismatched types")
-                .note_expected_found(&"type", &ty_hint, &format!("{}", cv.description()))
-                .span_label(e.span, &format!("expected '{}' type", ty_hint))
-                .emit();
-        };
+fn evaluate_disr_expr(ccx: &CrateCtxt, repr_ty: attr::IntType, body: hir::BodyId)
+                      -> Option<ty::Disr> {
+    let e = &ccx.tcx.hir.body(body).value;
+    debug!("disr expr, checking {}", ccx.tcx.hir.node_to_pretty_string(e.id));
+
+    let ty_hint = repr_ty.to_ty(ccx.tcx);
+    let print_err = |cv: ConstVal| {
+        struct_span_err!(ccx.tcx.sess, e.span, E0079, "mismatched types")
+            .note_expected_found(&"type", &ty_hint, &format!("{}", cv.description()))
+            .span_label(e.span, &format!("expected '{}' type", ty_hint))
+            .emit();
+    };
 
-        let hint = UncheckedExprHint(ty_hint);
-        match ConstContext::new(ccx.tcx, body).eval(e, hint) {
-            Ok(ConstVal::Integral(i)) => {
-                // FIXME: eval should return an error if the hint is wrong
-                match (repr_ty, i) {
-                    (attr::SignedInt(ast::IntTy::I8), ConstInt::I8(_)) |
-                    (attr::SignedInt(ast::IntTy::I16), ConstInt::I16(_)) |
-                    (attr::SignedInt(ast::IntTy::I32), ConstInt::I32(_)) |
-                    (attr::SignedInt(ast::IntTy::I64), ConstInt::I64(_)) |
-                    (attr::SignedInt(ast::IntTy::I128), ConstInt::I128(_)) |
-                    (attr::SignedInt(ast::IntTy::Is), ConstInt::Isize(_)) |
-                    (attr::UnsignedInt(ast::UintTy::U8), ConstInt::U8(_)) |
-                    (attr::UnsignedInt(ast::UintTy::U16), ConstInt::U16(_)) |
-                    (attr::UnsignedInt(ast::UintTy::U32), ConstInt::U32(_)) |
-                    (attr::UnsignedInt(ast::UintTy::U64), ConstInt::U64(_)) |
-                    (attr::UnsignedInt(ast::UintTy::U128), ConstInt::U128(_)) |
-                    (attr::UnsignedInt(ast::UintTy::Us), ConstInt::Usize(_)) => Some(i),
-                    (_, i) => {
-                        print_err(ConstVal::Integral(i));
-                        None
-                    },
-                }
-            },
-            Ok(cv) => {
-                print_err(cv);
-                None
-            },
-            // enum variant evaluation happens before the global constant check
-            // so we need to report the real error
-            Err(err) => {
-                let mut diag = report_const_eval_err(
-                    ccx.tcx, &err, e.span, "enum discriminant");
-                diag.emit();
-                None
+    let hint = UncheckedExprHint(ty_hint);
+    match ConstContext::new(ccx.tcx, body).eval(e, hint) {
+        Ok(ConstVal::Integral(i)) => {
+            // FIXME: eval should return an error if the hint is wrong
```

Comment:
It is a preexisting FIXME, it is saying that const-evaluation should fail if hint is incorrect so the check below shouldn’t be necessary. I should double-check whether it is not fixed yet.